File: src/physics/AABB.js
/**
 * Axis-Aligned Bounding Box utilities.
 * Provides AABB creation, intersection, and manipulation functions.
 * @module physics/AABB
 */

/**
 * @typedef {Object} AABB
 * @property {number} minX - Minimum X coordinate
 * @property {number} minY - Minimum Y coordinate
 * @property {number} minZ - Minimum Z coordinate
 * @property {number} maxX - Maximum X coordinate
 * @property {number} maxY - Maximum Y coordinate
 * @property {number} maxZ - Maximum Z coordinate
 */

/**
 * Create an AABB from center position and half-extents.
 * @param {number} x - Center X
 * @param {number} y - Center Y
 * @param {number} z - Center Z
 * @param {number} halfWidth - Half-width (X extent)
 * @param {number} halfHeight - Half-height (Y extent)
 * @param {number} halfDepth - Half-depth (Z extent)
 * @returns {AABB} The created AABB
 */
export function createAABB(x, y, z, halfWidth, halfHeight, halfDepth) {
    return {
        minX: x - halfWidth,
        minY: y - halfHeight,
        minZ: z - halfDepth,
        maxX: x + halfWidth,
        maxY: y + halfHeight,
        maxZ: z + halfDepth
    };
}

/**
 * Create an AABB from min/max coordinates directly.
 * @param {number} minX - Minimum X
 * @param {number} minY - Minimum Y
 * @param {number} minZ - Minimum Z
 * @param {number} maxX - Maximum X
 * @param {number} maxY - Maximum Y
 * @param {number} maxZ - Maximum Z
 * @returns {AABB} The created AABB
 */
export function createAABBFromMinMax(minX, minY, minZ, maxX, maxY, maxZ) {
    return { minX, minY, minZ, maxX, maxY, maxZ };
}

/**
 * Create an AABB for a block at integer coordinates.
 * Blocks occupy a 1x1x1 unit cube from (x, y, z) to (x+1, y+1, z+1).
 * @param {number} bx - Block X coordinate
 * @param {number} by - Block Y coordinate
 * @param {number} bz - Block Z coordinate
 * @returns {AABB} Block bounding box
 */
export function createBlockAABB(bx, by, bz) {
    return {
        minX: bx,
        minY: by,
        minZ: bz,
        maxX: bx + 1,
        maxY: by + 1,
        maxZ: bz + 1
    };
}

/**
 * Create an AABB for a player/entity from feet position and dimensions.
 * @param {number} x - Center X position
 * @param {number} y - Feet Y position (bottom of entity)
 * @param {number} z - Center Z position
 * @param {number} width - Entity width (X and Z)
 * @param {number} height - Entity height (Y)
 * @returns {AABB} Entity bounding box
 */
export function createEntityAABB(x, y, z, width, height) {
    const halfWidth = width / 2;
    return {
        minX: x - halfWidth,
        minY: y,
        minZ: z - halfWidth,
        maxX: x + halfWidth,
        maxY: y + height,
        maxZ: z + halfWidth
    };
}

/**
 * Check if two AABBs intersect.
 * Uses strict less-than for edge-to-edge touching (no overlap).
 * @param {AABB} a - First AABB
 * @param {AABB} b - Second AABB
 * @returns {boolean} True if boxes intersect
 */
export function intersectsAABB(a, b) {
    return (
        a.minX < b.maxX && a.maxX > b.minX &&
        a.minY < b.maxY && a.maxY > b.minY &&
        a.minZ < b.maxZ && a.maxZ > b.minZ
    );
}

/**
 * Check if a point is inside an AABB (inclusive bounds).
 * @param {AABB} box - The bounding box
 * @param {number} x - Point X
 * @param {number} y - Point Y
 * @param {number} z - Point Z
 * @returns {boolean} True if point is inside
 */
export function containsPoint(box, x, y, z) {
    return (
        x >= box.minX && x <= box.maxX &&
        y >= box.minY && y <= box.maxY &&
        z >= box.minZ && z <= box.maxZ
    );
}

/**
 * Check if an AABB fully contains another AABB.
 * @param {AABB} outer - The containing box
 * @param {AABB} inner - The box to check
 * @returns {boolean} True if outer fully contains inner
 */
export function containsAABB(outer, inner) {
    return (
        inner.minX >= outer.minX && inner.maxX <= outer.maxX &&
        inner.minY >= outer.minY && inner.maxY <= outer.maxY &&
        inner.minZ >= outer.minZ && inner.maxZ <= outer.maxZ
    );
}

/**
 * Expand an AABB by a given amount in all directions.
 * @param {AABB} box - The bounding box
 * @param {number} amount - Amount to expand (can be negative to shrink)
 * @returns {AABB} New expanded AABB
 */
export function expandAABB(box, amount) {
    return {
        minX: box.minX - amount,
        minY: box.minY - amount,
        minZ: box.minZ - amount,
        maxX: box.maxX + amount,
        maxY: box.maxY + amount,
        maxZ: box.maxZ + amount
    };
}

/**
 * Expand an AABB by different amounts per axis.
 * @param {AABB} box - The bounding box
 * @param {number} amountX - X expansion
 * @param {number} amountY - Y expansion
 * @param {number} amountZ - Z expansion
 * @returns {AABB} New expanded AABB
 */
export function expandAABBAxis(box, amountX, amountY, amountZ) {
    return {
        minX: box.minX - amountX,
        minY: box.minY - amountY,
        minZ: box.minZ - amountZ,
        maxX: box.maxX + amountX,
        maxY: box.maxY + amountY,
        maxZ: box.maxZ + amountZ
    };
}

/**
 * Translate an AABB by an offset.
 * @param {AABB} box - The bounding box
 * @param {number} dx - X offset
 * @param {number} dy - Y offset
 * @param {number} dz - Z offset
 * @returns {AABB} New translated AABB
 */
export function translateAABB(box, dx, dy, dz) {
    return {
        minX: box.minX + dx,
        minY: box.minY + dy,
        minZ: box.minZ + dz,
        maxX: box.maxX + dx,
        maxY: box.maxY + dy,
        maxZ: box.maxZ + dz
    };
}

/**
 * Get the center of an AABB.
 * @param {AABB} box - The bounding box
 * @returns {{x: number, y: number, z: number}} Center point
 */
export function getAABBCenter(box) {
    return {
        x: (box.minX + box.maxX) / 2,
        y: (box.minY + box.maxY) / 2,
        z: (box.minZ + box.maxZ) / 2
    };
}

/**
 * Get the dimensions of an AABB.
 * @param {AABB} box - The bounding box
 * @returns {{width: number, height: number, depth: number}} Dimensions
 */
export function getAABBSize(box) {
    return {
        width: box.maxX - box.minX,
        height: box.maxY - box.minY,
        depth: box.maxZ - box.minZ
    };
}

/**
 * Get the volume of an AABB.
 * @param {AABB} box - The bounding box
 * @returns {number} Volume
 */
export function getAABBVolume(box) {
    return (box.maxX - box.minX) * (box.maxY - box.minY) * (box.maxZ - box.minZ);
}

/**
 * Compute the union of two AABBs (smallest box containing both).
 * @param {AABB} a - First AABB
 * @param {AABB} b - Second AABB
 * @returns {AABB} Union AABB
 */
export function unionAABB(a, b) {
    return {
        minX: Math.min(a.minX, b.minX),
        minY: Math.min(a.minY, b.minY),
        minZ: Math.min(a.minZ, b.minZ),
        maxX: Math.max(a.maxX, b.maxX),
        maxY: Math.max(a.maxY, b.maxY),
        maxZ: Math.max(a.maxZ, b.maxZ)
    };
}

/**
 * Compute the intersection of two AABBs.
 * Returns null if boxes don't intersect.
 * @param {AABB} a - First AABB
 * @param {AABB} b - Second AABB
 * @returns {AABB|null} Intersection AABB or null
 */
export function intersectionAABB(a, b) {
    const minX = Math.max(a.minX, b.minX);
    const minY = Math.max(a.minY, b.minY);
    const minZ = Math.max(a.minZ, b.minZ);
    const maxX = Math.min(a.maxX, b.maxX);
    const maxY = Math.min(a.maxY, b.maxY);
    const maxZ = Math.min(a.maxZ, b.maxZ);

    if (minX >= maxX || minY >= maxY || minZ >= maxZ) {
        return null;
    }

    return { minX, minY, minZ, maxX, maxY, maxZ };
}

/**
 * Clone an AABB.
 * @param {AABB} box - The bounding box
 * @returns {AABB} Copy of the AABB
 */
export function cloneAABB(box) {
    return {
        minX: box.minX,
        minY: box.minY,
        minZ: box.minZ,
        maxX: box.maxX,
        maxY: box.maxY,
        maxZ: box.maxZ
    };
}

/**
 * Copy values from one AABB to another (mutates target).
 * @param {AABB} target - Target AABB to modify
 * @param {AABB} source - Source AABB to copy from
 * @returns {AABB} The target AABB
 */
export function copyAABB(target, source) {
    target.minX = source.minX;
    target.minY = source.minY;
    target.minZ = source.minZ;
    target.maxX = source.maxX;
    target.maxY = source.maxY;
    target.maxZ = source.maxZ;
    return target;
}

/**
 * Set AABB values (mutates box).
 * @param {AABB} box - The bounding box to modify
 * @param {number} minX - Minimum X
 * @param {number} minY - Minimum Y
 * @param {number} minZ - Minimum Z
 * @param {number} maxX - Maximum X
 * @param {number} maxY - Maximum Y
 * @param {number} maxZ - Maximum Z
 * @returns {AABB} The modified AABB
 */
export function setAABB(box, minX, minY, minZ, maxX, maxY, maxZ) {
    box.minX = minX;
    box.minY = minY;
    box.minZ = minZ;
    box.maxX = maxX;
    box.maxY = maxY;
    box.maxZ = maxZ;
    return box;
}

File: voxel_engine/engine/physics/aabb.py
"""
AABB collision primitives using NumPy arrays.

Axis-Aligned Bounding Boxes are represented as (6,) float64 arrays:
[min_x, min_y, min_z, max_x, max_y, max_z]

This representation allows efficient NumPy operations and Numba JIT.

Usage:
    from voxel_engine.engine.physics.aabb import make_aabb, aabb_intersects

    box_a = make_aabb(np.array([0, 0, 0]), np.array([1, 1, 1]))
    box_b = make_aabb(np.array([0.5, 0, 0]), np.array([1.5, 1, 1]))
    if aabb_intersects(box_a, box_b):
        print("Collision!")
"""

import numpy as np
from numpy.typing import NDArray


def make_aabb(
    min_corner: NDArray[np.float64],
    max_corner: NDArray[np.float64]
) -> NDArray[np.float64]:
    """
    Create AABB as (6,) array from min/max corners.

    Args:
        min_corner: (3,) array of minimum X, Y, Z.
        max_corner: (3,) array of maximum X, Y, Z.

    Returns:
        (6,) array: [min_x, min_y, min_z, max_x, max_y, max_z].
    """
    return np.array([
        min_corner[0], min_corner[1], min_corner[2],
        max_corner[0], max_corner[1], max_corner[2]
    ], dtype=np.float64)


def aabb_from_center(
    center: NDArray[np.float64],
    half_extents: NDArray[np.float64]
) -> NDArray[np.float64]:
    """
    Create AABB from center point and half-widths.

    Args:
        center: (3,) array of center X, Y, Z.
        half_extents: (3,) array of half-widths in X, Y, Z.

    Returns:
        (6,) AABB array.
    """
    return make_aabb(center - half_extents, center + half_extents)


def aabb_intersects(
    a: NDArray[np.float64],
    b: NDArray[np.float64]
) -> bool:
    """
    Check if two AABBs overlap.

    Uses separating axis theorem for axis-aligned boxes.

    Args:
        a: First AABB (6,) array.
        b: Second AABB (6,) array.

    Returns:
        True if boxes overlap, False otherwise.
    """
    # Check all three axes for separation
    # Overlap requires: a.min <= b.max AND a.max >= b.min on all axes
    return (
        a[0] <= b[3] and a[3] >= b[0] and  # X overlap
        a[1] <= b[4] and a[4] >= b[1] and  # Y overlap
        a[2] <= b[5] and a[5] >= b[2]      # Z overlap
    )


def aabb_contains_point(
    aabb: NDArray[np.float64],
    point: NDArray[np.float64]
) -> bool:
    """
    Check if point is inside AABB (inclusive).

    Args:
        aabb: AABB (6,) array.
        point: (3,) array of point coordinates.

    Returns:
        True if point is inside or on boundary.
    """
    return (
        aabb[0] <= point[0] <= aabb[3] and
        aabb[1] <= point[1] <= aabb[4] and
        aabb[2] <= point[2] <= aabb[5]
    )


def player_aabb(
    position: NDArray[np.float64],
    width: float,
    height: float
) -> NDArray[np.float64]:
    """
    Create player AABB from feet position.

    Player position is at the feet (bottom center).
    The AABB extends from feet up to head height.

    Args:
        position: (3,) array of player feet position.
        width: Player width (X and Z extent).
        height: Player height (Y extent from feet).

    Returns:
        (6,) AABB array.
    """
    half_width = width / 2.0
    return make_aabb(
        np.array([
            position[0] - half_width,
            position[1],
            position[2] - half_width
        ], dtype=np.float64),
        np.array([
            position[0] + half_width,
            position[1] + height,
            position[2] + half_width
        ], dtype=np.float64)
    )


def swept_aabb(
    aabb: NDArray[np.float64],
    velocity: NDArray[np.float64]
) -> NDArray[np.float64]:
    """
    Expand AABB to encompass entire movement path.

    Creates a bounding box that contains the original AABB
    at both its start and end positions after velocity is applied.
    Useful for broad-phase collision detection.

    Args:
        aabb: Original AABB (6,) array.
        velocity: (3,) movement velocity for this frame/tick.

    Returns:
        Expanded AABB (6,) array.
    """
    expanded = aabb.copy()
    for i in range(3):
        if velocity[i] < 0:
            expanded[i] += velocity[i]  # Expand min
        else:
            expanded[i + 3] += velocity[i]  # Expand max
    return expanded


def aabb_get_center(aabb: NDArray[np.float64]) -> NDArray[np.float64]:
    """
    Get center point of AABB.

    Args:
        aabb: AABB (6,) array.

    Returns:
        (3,) array of center coordinates.
    """
    return np.array([
        (aabb[0] + aabb[3]) / 2.0,
        (aabb[1] + aabb[4]) / 2.0,
        (aabb[2] + aabb[5]) / 2.0
    ], dtype=np.float64)


def aabb_get_size(aabb: NDArray[np.float64]) -> NDArray[np.float64]:
    """
    Get size (dimensions) of AABB.

    Args:
        aabb: AABB (6,) array.

    Returns:
        (3,) array of width, height, depth.
    """
    return np.array([
        aabb[3] - aabb[0],
        aabb[4] - aabb[1],
        aabb[5] - aabb[2]
    ], dtype=np.float64)
